bst.py

```diff
@@ -0,0 +1,179 @@
+import random
+from timeit import default_timer
+
+
+class Node(object):
+        def __init__(self, val, parent=None, left=None, right=None):
+            self.val = val
+            self.parent = parent
+            self.left = left
+            self.right = right
+
+        def __repr__(self):
+            return ("Node({self.val}, parent={self.parent})".format(self=self))
+
+        def get_dot(self):
+            """return the tree with root 'self' as a dot graph for visualization"""
+            return "digraph G{\n%s}" % ("" if self.val is None else (
+                "\t%s;\n%s\n" % (
+                    self.val,
+                    "\n".join(self._get_dot())
+                )
+            ))
+
+        def _get_dot(self):
+            """recursively prepare a dot graph entry for this node."""
+            if self.left is not None:
+                yield "\t%s -> %s;" % (self.val, self.left.val)
+                for i in self.left._get_dot():
+                    yield i
+            elif self.right is not None:
+                r = random.randint(0, 1e9)
+                yield "\tnull%s [shape=point];" % r
+                yield "\t%s -> null%s;" % (self.val, r)
+            if self.right is not None:
+                yield "\t%s -> %s;" % (self.val, self.right.val)
+                for i in self.right._get_dot():
+                    yield i
+            elif self.left is not None:
+                r = random.randint(0, 1e9)
+                yield "\tnull%s [shape=point];" % r
+                yield "\t%s -> null%s;" % (self.val, r)
+
+
+class BST(object):
+    def __init__(self, val=None):
+        self._size = 0
+        self.root = None
+        if val is not None:
+            self.insert(val)
+        else:
+            self.root = None
+
+    def insert(self, val):
+        """Insert val into BST if not present. Always returns None."""
+        if self.root is None:
+            self.root = Node(val)
+            self._size += 1
+        else:
+            new_parent = self._find(val)
+            if new_parent.val > val:
+                new_parent.left = Node(val, new_parent)
```

Comment:
What if it were possible to assign a value to a node's `left` or `right` property and have that automatically construct a `Node` and set that parent?  So that you could do this:

```python
new_parent.left = val
assert new_parent.left.parent is new_parent
```